apps/web/src/hooks/useBalances.ts

```diff
@@ -1,24 +1,70 @@
+import { getCounterfactualBalance } from '@/features/counterfactual/utils'
+import { useWeb3 } from '@/hooks/wallets/web3'
 import { useMemo } from 'react'
-import isEqual from 'lodash/isEqual'
 import { type SafeBalanceResponse } from '@safe-global/safe-gateway-typescript-sdk'
 import { useAppSelector } from '@/store'
-import { initialBalancesState, selectBalances } from '@/store/balancesSlice'
+import useAsync from '@/hooks/useAsync'
+import { selectCurrency, selectSettings, TOKEN_LISTS } from '@/store/settingsSlice'
+import { useCurrentChain } from '@/hooks/useChains'
+import { FEATURES, hasFeature } from '@/utils/chains'
+import useSafeInfo from '@/hooks/useSafeInfo'
+import { useBalancesGetBalancesV1Query } from '@safe-global/store/gateway/AUTO_GENERATED/balances'
+import { ExtendedSafeInfo } from '@/store/safeInfoSlice'
 
-const useBalances = (): {
-  balances: SafeBalanceResponse
-  loading: boolean
-  error?: string
-} => {
-  const state = useAppSelector(selectBalances, isEqual)
-  const { data, error, loading } = state
+export const useTokenListSetting = (): boolean | undefined => {
+  const chain = useCurrentChain()
+  const settings = useAppSelector(selectSettings)
+
+  const isTrustedTokenList = useMemo(() => {
+    if (settings.tokenList === TOKEN_LISTS.ALL) return false
+    return chain ? hasFeature(chain, FEATURES.DEFAULT_TOKENLIST) : undefined
+  }, [chain, settings.tokenList])
+
+  return isTrustedTokenList
+}
+
+const useCounterfactualBalances = (safe: ExtendedSafeInfo) => {
+  const web3 = useWeb3()
+  const chain = useCurrentChain()
+  const safeAddress = safe.address.value
+  const isCounterfactual = !safe.deployed
+
+  const [data] = useAsync<SafeBalanceResponse | undefined>(() => {
+    if (!chain || !web3 || !isCounterfactual) return
+    return getCounterfactualBalance(safeAddress, web3, chain)
+  }, [chain, safeAddress, web3, isCounterfactual])
+
+  return data
+}
+
+const useBalances = () => {
+  const currency = useAppSelector(selectCurrency)
+  const isTrustedTokenList = useTokenListSetting()
+  const { safe, safeAddress, safeLoaded } = useSafeInfo()
+  const isReady = safeLoaded && isTrustedTokenList !== undefined
+
+  const { data, isLoading, error } = useBalancesGetBalancesV1Query(
```

Comment:
Are we still polling the balances?

Reply:
Part two of this PR is going to be to replace balances polling with a Safe Overview fetching because we actually only need the fiat total, not all the balances.